### ISSUE TITLE
Error: Solver failed

### ISSUE THREAD

**Issue by Sabor117** [privefl/bigsnpr#107]
Hi Florian,

Apologies for the third issue in as many weeks, but I've stumbled across another problem with using LDpred2 that I could use some help with.

When running snp_ldpred2_auto() I have been receiving the error:
```
Error in sp_solve_sym_eigen(A, b, add_to_diag, tol, maxiter) : 
  Solver failed.
```
After examining the code again, we thought that the issue might have been with either the number of iterations, or potentially the tolerance argument in bigsparser::sp_solve_sym(), as we thought the basal level of 1e-10 may have been too stringent. So I modified our local version of the code so that you I could change the tol argument of sp_solve_sym() from the level of snp_ldpred2_auto(). This resulted in me trying different variations all the way to:
```
multi_auto = snp_ldpred2_auto(corr, df_beta, h2_init = h2_est,
                           vec_p_init = seq_log(1e-4, 0.9, nb_cores()), ncores = nb_cores(),
                           num_iter = 50000, in_tolerance = 1e-3)
```
And even this resulted in the same error.

I've basically been a bit stumped by this and couldn't find any other documentation about the error, so I was hoping you might be able to advise if there was any way of fixing this (or whether this basically meant the method would not work for this trait).

**Comment by privefl** [comment]
I believe it means that the LD matrix is not invertible, even after adding `m / (n * h2)` on the diagonal (cf. https://github.com/privefl/bigsnpr/blob/master/R/LDpred2.R#L129-L130).
What values do you have for these 3 parameters?

**Comment by Sabor117** [comment]
So, assuming I'm understanding this correctly and `m / (n * h2)` from your answer is equivalent to `add_to_diag = ncol(corr) / (h2_init * N)` from the highlighted code (and not corr or beta_hat), then my values are as follows:
```
m (ncol(corr)) = 721
h2 (h2_init) = 0.3870018
n (N) = [See below]
```
So for `N` I have a list of 721 sample sizes for each (from `df_beta$n_eff`). The values range from 956 up to ~31563.

Is N supposed to be one single value?

**Comment by privefl** [comment]
No, N is not supposed to be a single value but usually people remove values lower than 0.7 × max(N). I did not investigated this further.
However, M seems small. 

And are you sure the LD matrix is okay?

**Comment by Sabor117** [comment]
Yes, I'm actually running a number of instances with a pretty small number of SNPs, often around that number. We have a pretty specific selection criteria.

I'll look into what happens with removing the lower values for N. But what do you mean by the LD matrix? Do you mean `corr` (from your tutorials)?

Looking at `corr0` it seems to be okay, no missing or NA values.

**Comment by privefl** [comment]
Yes. Ok. Keep me updated.

**Comment by bvilhjal** [comment]
Could it be that the number of samples being used to calculate the LD matrix is small?

**Comment by Sabor117** [comment]
> Could it be that the number of samples being used to calculate the LD matrix is small?

For this, do you mean the h2 calculation? Because if so, then no.

Because of some other issues (which I think I reported previously) I actually have been using two separate datasets. I have the dataset I am attempting to calculate the PRS for with LDpred2, and then for calculating h2 I actually have been using a reference of 10k individuals from UKBB.

**Comment by privefl** [comment]
But you're not using these 10K individuals for computing the LD matrix used in LDpred2?

**Comment by bvilhjal** [comment]
I would think you only need two datasets for fitting the model.  The sum stats and an LD reference.  Using 10K (relatively unrelated) indivs from UKB sounds reasonable as an LD reference, i.e. those would be ones you would use to calculate `corr`.  Then for estimating h2, and running LDpred2 you'd also need the summary stats.

**Comment by privefl** [comment]
Any update on this?

**Comment by privefl** [comment]
Thank you for using LDpred2. Please note that we now recommend running LDpred2 **genome-wide** instead of per chromosome. The paper (preprint) and tutorial have been updated.

**Comment by Sabor117** [comment]
Hi Florian, apologies for the delay in getting back to you about this older issue. For posterity I thought I would briefly mention how I fixed the issue.

I believe you hit the nail on the head when you said:

> But you're not using these 10K individuals for computing the LD matrix used in LDpred2?

When using ```snp_ldpred2_auto``` (below) I was only using the 10K individuals for the h2 estimate. I was however, using my test cohort for the correlation matrix (corr).

    multi_auto <- snp_ldpred2_auto(corr, df_beta, h2_init = h2_est,
                               vec_p_init = seq_log(1e-4, 0.9, length.out = NCORES),
                               ncores = NCORES)

In hindsight, in order to create the polygenic weights with ```snp_ldpred2_auto``` it didn't need my test cohort, so this was a mistake. I'm not sure why exactly it was failing, but your suggestion that it was too small a sample size seems plausible (it is only 1K individuals).

Tl;dr: fixing this involved using the 10K reference panel for both corr and h2_est.

**Comment by privefl** [comment]
Great to know that you fixed this.

Also note that we now provide an LD reference based on 360K UKBB European individuals and HapMap3 variants (see new version of the preprint), in case some researchers don't have access to enough samples for computing the LD ref.

**Comment by Sabor117** [comment]
Is this a reference panel in bgen format? I have access to the full UKBB data, but used a subset of 10K individuals primarily to save computational time. I am now wondering if it makes more sense to use the whole dataset, if that is what you recommend.

Also, may I ask about this:

> Thank you for using LDpred2. Please note that we now recommend running LDpred2 **genome-wide** instead of per chromosome. The paper (preprint) and tutorial have been updated.

Is this because of computational reasons or biological reasons?

**Comment by privefl** [comment]
10K is okay.

It is the LD matrices, basically what you get when running `snp_cor()`.

We were recommending per chromosome for computational and possible biological reasons, but running it genome-wide just provides better predictive performance, so..
